### PR TITLE
✨ Improve GitHub SARIF generation

### PR DESCRIPTION
**Pull request by laurentsimon**
* **Please check if the PR fulfills these requirements**
- [] Tests for the changes have been added (for bug fixes / features)
- [X] PR title follows the guidelines defined in  https://github.com/ossf/scorecard/blob/main/CONTRIBUTING.md#pr-process

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
feature

* **What is the current behavior?** (You can also link to an open issue here)
https://github.com/ossf/scorecard/issues/1074

* **What is the new behavior (if this is a feature change)?**
https://github.com/ossf/scorecard/issues/1074

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
no

* **Other information**:

### CODE REVIEW

File: pkg/sarif.go [hunk 1]
```diff
@@ -382,40 +349,55 @@ func createSARIFHeader(url, category, name, version, commit string, t time.Time)
...
+	r := ""
+	for _, s := range remediation {
+		r += fmt.Sprintf("- %s\n\n", s)
+	}
+	return r
+}
+
+func generateMarkdownText(longDesc, risk string, remediation []string) string {
+	rSev := fmt.Sprintf("**Severity**: %s", risk)
+	rDesc := fmt.Sprintf("**Details**: \n%s", longDesc)
```
Comment (oliverchang) [inline pkg/sarif.go:359]: nit: no need for trailing space after : ?

File: pkg/sarif.go [hunk 2]
```diff
@@ -382,40 +349,55 @@ func createSARIFHeader(url, category, name, version, commit string, t time.Time)
 	}
 }
 
-func createSARIFRule(checkName, checkID, descURL, longDesc, shortDesc string,
+func generateRecommentation(remediation []string) string {
```
Comment (oliverchang) [inline pkg/sarif.go:349]: nit: spelling of "recommen**d**ation"

File: pkg/sarif.go [hunk 3]
```diff
@@ -382,40 +349,55 @@ func createSARIFHeader(url, category, name, version, commit string, t time.Time)
...
+	r := ""
+	for _, s := range remediation {
+		r += fmt.Sprintf("- %s\n\n", s)
```
Comment (oliverchang) [inline pkg/sarif.go:352]: Nit: Do we need the two newlines here, or is one enough?

**Review by oliverchang — APPROVED** [review]

File: pkg/sarif.go [hunk 4]
```diff
@@ -84,9 +80,12 @@ type defaultConfig struct {
 	Level string `json:"level"`
 }
 
+//nolint:govet
```
Comment (azeemshaikh38) [inline pkg/sarif.go:83]: Should be fixable by putting `Tags` as the last field.

File: pkg/sarif.go [hunk 5]
```diff
@@ -156,6 +155,44 @@ func maxOffset(x, y int) int {
 	return x
 }
 
+func calculateSeverityLevel(risk string) string {
+	// nolint
```
Comment (azeemshaikh38) [inline pkg/sarif.go:159]: Specify the linter here.

File: pkg/sarif.go [hunk 6]
```diff
@@ -156,6 +155,44 @@ func maxOffset(x, y int) int {
...
+	// https://docs.github.com/en/code-security/code-scanning/integrating-with-code-scanning/sarif-support-for-code-scanning#reportingdescriptor-object.
+	// "over 9.0 is critical, 7.0 to 8.9 is high, 4.0 to 6.9 is medium and 3.9 or less is low".
+	switch risk {
+	case "Critical":
+		return "9.0"
+	case "High":
+		return "7.0"
+	case "Medium":
+		return "4.0"
+	// Low
```
Comment (azeemshaikh38) [inline pkg/sarif.go:169]: Maybe have a case for "Low" too and panic on `default`

File: pkg/sarif.go [hunk 7]
```diff
@@ -156,6 +155,44 @@ func maxOffset(x, y int) int {
...
+	default:
+		return "1.0"
+	}
+}
+
+func generateProblemSeverity(risk string) string {
+	// nolint
```
Comment (azeemshaikh38) [inline pkg/sarif.go:176]: Specify the failing linter

**Review by azeemshaikh38 — APPROVED** [review]

File: pkg/sarif.go (hunk 2 above)
Comment (laurentsimon) [reply pkg/sarif.go]: even worse.. this was supposed to be "remediation" :-). Fixed.

**Pull request merged**